### PR TITLE
Add re-fetch query watchers

### LINKED ISSUE #452: Introduce RefetchQuerries for mutations
Javascript documentation: http://dev.apollodata.com/core/apollo-client-api.html#ApolloClient.mutate

Use case: There is post feed, and a detail page where you can create a post. When we create the post, there may be no way for the normalized cache to know that the new post belongs to the feed query. One option is to directly write into the store, but an even simpler (albeit less efficient) is to specify "RefetchQueries" for the mutation.

Ideally, we will want to batch all the qurries in the refetch query into one POST call.

### PR DESCRIPTION
**Pull request by sav007**
Re-fetch query watchers after mutation completed
Refactor ApolloTracker.

Closes #452

### CODE REVIEW

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java [hunk 1]
```diff
@@ -210,61 +211,69 @@ private RealApolloCall(Builder<T> builder) {
...
-        if (queryFetcher.isPresent()) {
-          queryFetcher.get().refetchAsync(new QueryFetcher.OnFetchCompleteCallback() {
-            @Override public void onFetchComplete() {
-              //noinspection unchecked
-              responseCallback.onResponse(response.parsedResponse.get());
-            }
-          });
-        } else { //noinspection unchecked
-          responseCallback.onResponse(response.parsedResponse.get());
+  private void refetchQueryWatchers() {
```
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: Although strange, should this be `refetchRefetchQueryWatchers`?

I'm a little confused by this. We should never need to explicity call refetch on a query watcher, as the point of a query watcher is to refetch itself when a watched key changes.

We only want to refetch the expliciting set refetchQuerires.

Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: Maybe this goes back to the discussion we had a while ago, I've gotten a little lost, but do we force user's to be watching a query in able to refetch it?

I've also envisioned them to be separate concepts -- you can watch whatever you want, and whenever a dependent key changes you get notified. And when you mutate, you can pass a set of real query objects to refetch -- these will have no direct listeners, _but_ if you are watching somewhere, the normal watcher flow will trigger and notify.

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java [hunk 2]
```diff
@@ -210,61 +211,69 @@ private RealApolloCall(Builder<T> builder) {
...
+            return;
+          }
+
+          refetchQueryWatchers();
+
+          if (queryFetcher.isPresent()) {
+            queryFetcher.get().refetchAsync(new QueryFetcher.OnFetchCompleteCallback() {
+              @Override public void onFetchComplete() {
+                //noinspection unchecked
+                originalCallback.onResponse(response.parsedResponse.get());
```
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: Would it be safe to not block the original call on the refetch queries? I can see the argument for both ways, ideally if it blocked or not would be user configurable.

Additionally, the actually refetch routine seems to be serially executed. Maybe we should investigate just dumping all the refetch querries into one graphql query? Although is may complicate query whitelisting.

**Comment by sav007** [comment]
@BenSchwab 

> Also, do we need to guard against an infinite loop of refetch queries? (Two refetch queries reference each other and ping pong back and forth refetching)

That should not be the case, as refetch queries names can be set only for `ApolloMutationCall`. At the same time you can't watch `ApolloMutationCall` only query can be watched. So we are safe here.

> Would it be safe to not block the original call on the refetch queries? I can see the argument for both ways, ideally if it blocked or not would be user configurable.

Yeah I thought about this and don't have strong opinion on that so not sure what the right way.

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java (hunk 1 above)
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: Sorry what do you mean?

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java (hunk 2 above)
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: > Maybe we should investigate just dumping all the refetch querries into one graphql query? 

Yeah that will be awesome, we def need to explore this, I see though as it won't be trivial. At the same time we can explore overall merging queries (smth like throttling ) into one call.

**Comment by stubailo** [comment]
> Maybe we should investigate just dumping all the refetch queries into one graphql query? Although is may complicate query whitelisting.

A better approach would be to use transport batching probably (sending multiple documents/query IDs in one request), that way you can still whitelist, but they can be sent in one request and batched on the server.

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java (hunk 1 above)
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java]: 'refetchQueryWatchers` sort of makes it seem like we will be refetching all query watchers - but really we are refetching the set of "refetch" query watchers, right?

It's an internal method, so not a big deal at all.

**Comment by sav007** [comment]
@digitalbuddha any comments?

**Pull request merged**